gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/iap/IAPAuthenticationFilter.groovy

```diff
@@ -0,0 +1,204 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.gate.security.iap
+
+import com.google.common.base.Preconditions
+import com.netflix.spinnaker.gate.services.PermissionService
+import com.netflix.spinnaker.gate.services.internal.Front50Service
+import com.netflix.spinnaker.security.User
+import com.nimbusds.jose.JWSHeader
+import com.nimbusds.jose.JWSVerifier
+import com.nimbusds.jose.crypto.ECDSAVerifier
+import com.nimbusds.jose.jwk.ECKey
+import com.nimbusds.jose.jwk.JWK
+import com.nimbusds.jose.jwk.JWKSet
+import com.nimbusds.jwt.JWTClaimsSet
+import com.nimbusds.jwt.SignedJWT
+import groovy.util.logging.Slf4j
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.security.core.context.SecurityContextHolder
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
+import retrofit.RetrofitError
+
+import javax.servlet.Filter
+import javax.servlet.FilterChain
+import javax.servlet.FilterConfig
+import javax.servlet.ServletException
+import javax.servlet.ServletRequest
+import javax.servlet.ServletResponse
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+import java.security.interfaces.ECPublicKey
+
+/***
+ * This filter verifies the request header from Cloud IAP containing a JWT token, after the user
+ * has been authenticated and authorized by IAP via Google OAuth2.0 and IAP's authorization service.
+ * The user email from the payload used to create the Spinnaker user.
+ */
+@Slf4j
+class IAPAuthenticationFilter implements Filter {
+
+  IAPConfig.IAPSecurityConfigProperties configProperties
+
+  PermissionService permissionService
+
+  Front50Service front50Service
+
+  private static final String signatureAttribute = "JWTSignature"
+
+  private final Map<String, JWK> keyCache = new HashMap<>()
+  private final Logger logger = LoggerFactory.getLogger(IAPAuthenticationFilter.class)
+
+  IAPAuthenticationFilter(
+    IAPConfig.IAPSecurityConfigProperties configProperties,
+    PermissionService permissionService,
+    Front50Service front50Service) {
+    this.configProperties = configProperties
+    this.permissionService = permissionService
+    this.front50Service = front50Service
+  }
+
+  @Override
+  void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+    HttpServletRequest req = (HttpServletRequest) request
+    HttpServletResponse res = (HttpServletResponse) response
+
+
```

Comment:
nit: extra line here and various places below

Reply:
Removed.